### DISCUSSION TITLE
Is it possible to use terrain data in a custom style layer to shade the map?

### DISCUSSION THREAD

**Discussion by larsmaxfield** [maplibre/maplibre-gl-js#4133] [Q&A]
Is possible to use terrain data in a [custom style layer](https://maplibre.org/maplibre-gl-js/docs/examples/custom-style-layer/) to color the map based on elevation?

My goal is to have a custom layer which achieves this:

![image](https://github.com/maplibre/maplibre-gl-js/assets/83759569/f6e398e9-78ea-4a6a-8067-ce366524e8ce)

I achieved that^ by editing the existing shaders in the hillshading layer because they already have methods to calculate the elevation and shade the terrain.

I'd like to now properly implement this elevation coloring using the API.

[Custom layers](https://maplibre.org/maplibre-gl-js/docs/examples/custom-style-layer/) appear to be the way to do this because I can render custom WebGL content in them. However, I have not found a way forward which allows me to access terrain data and therefore determine elevation — that is why I'm asking here if it is even possible to do so.

**Comment by HarelM** [comment]
Terrain data can be fetched from the tiles, source cache, and stuff like that.
Custom layer can use the map object which can use some more "internal" stuff.
In theory, I think you can use the terrain-rgb tiles as "regular" raster tiles and simply use the terrain to make the 3D effect.
But I might not be correct here.
You can also use `addProtocol` to manipulate incoming data to achieve raster coloring.
But you'll need to play around with stuff to get to what you need.
I am looking for a more straight forward way to incorporate shaders into the code in terms of hooks, but I don't have a good enough understanding to actually design such hooks, so any help would be appreciated.

**Reply by larsmaxfield** [reply]
Thanks for the tips — and thanks for moving to discussions.

I'm in a similar spot not having a good enough understanding to jump in. The `addProtocol` mechanic I see is used to [add contour lines](https://maplibre.org/maplibre-gl-js/docs/examples/contour-lines/). Will take a look. But indeed a solution using shaders (from what I know) would be the most efficient.

**Reply by acalcutt** [reply]
This is an example I made using addProtocol to make color relief based on terrainrgb using rainbow.js
https://stackblitz.com/edit/web-platform-g15hva?file=index.html . 

my example returns back a image that gets used as raster tiles, but the maplibre-contour plugin example you pointed out works similar but returns vector tiles
https://github.com/onthegomap/maplibre-contour?tab=readme-ov-file#how-it-works

**Reply by acalcutt** [reply]
It would be great to see some faster build in support for this though, like is in the discussion for contour 
https://github.com/maplibre/maplibre-style-spec/issues/583

**Reply by larsmaxfield** [reply]
@acalcutt awesome! Quite a straightforward solution, and reminds me of [leaflet-topography](https://github.com/slutske22/leaflet-topography).

Built-in would be slick. I've started to play around with making an elevation source/layer as a way to learn how hillshading is implemented.

**Reply by acalcutt** [reply]
The visual part of it was based on leaflet examples from a blog article. Unfortunatly looking back I am not able to find the article I had read when making that example. It does look similar to that leaflet-topology though.

The addProtocol part actually came from the examples at the bottom of https://maplibre.org/news/2022-06-10-mierune-announcement/ 
https://jsbin.com/vonixulezo/edit?html,output 
when I saw their example it really helped adapt the leaflet terrain coloring examples I was looking at, since the code was visually similar.

**Comment by zouyaoji** [comment]
@larsmaxfield 
Hello, similarly, I've extended a custom layer based on the CustomLayerInterface, generating textures through WebGL, and rendering them using the layer. However, I'm unsure how to overlay them onto terrain. I'd like to discuss and seek your advice on possible solutions, if you're available.

**Reply by larsmaxfield** [reply]
Hi @zouyaoji, cool to see you're also working towards a more advanced custom layer. 

Unfortunately I don't have much advice to give — I'm still working on extending/reverse-engineering the hillshade layer in a robust way and learning how things work with MapLibre and web programming in general. I'll let you know once I've got my fork ready where we can collaborate more effectively.

In the meantime there is something I _think_ could help answer how to overlay on terrain: in `renderLayer` of `src/render/painter.ts` there are calls to `drawHillshade` and `drawCustom`:

```ts
    renderLayer(painter: Painter, sourceCache: SourceCache, layer: StyleLayer, coords: Array<OverscaledTileID>) {
        ...
        switch (layer.type) {
            ...
            case 'hillshade':
                drawHillshade(painter, sourceCache, layer as any, coords);
                break;
            ...
            case 'custom':
                drawCustom(painter, sourceCache, layer as any);
                break;
        }
    }
```

See how `drawCustom` doesn't include `coords` while `drawHillshade` does? Hillshade calls the `dem` of tiles for drawing, so that could be key to drawing on terrain. (My decision to reverse engineer the hillshade layer instead of custom layer was partly made because it indeed already achieves overlaying on the terrain.)

**Reply by larsmaxfield** [reply]
*Looking further to it, the `draw` calls in `draw_hillshade.ts`, `draw_circle.ts`, and other 3D-terrain-overlaying drawers use a `terrainData` variable. In `draw_custom.ts` this terrain doesn't seem to be used.
